### PR TITLE
Sanitize strings that would lead to uneaven quotes in line

### PR DESCRIPTION
**Pull request by WalternativE**
- Adds two tests based on failing formatting
  behavior having hash directives
- Documentation of triage session in Fantomas & Friends

### CODE REVIEW

**Comment by nojaf** [comment]
@Smaug123 we sorta went with a dirty hack to solve the problem. I'm going to revisit the solution and try and come up with something better ;)

**Comment by nojaf** [comment]
So I've refactored the code a bit and tried to come up with a better way of detecting all the defines inside a raw string.
The algorithm is rather primitive were all characters in the string are processed while being aware in what kind of code block you are currently in (string, multiline code comment)

Lastly, I refactored the code so that the define detection dance is only performed once.

File: .fantomasignore [hunk 1]
```diff
@@ -0,0 +1 @@
+TokenParser.fs
```
Comment (Smaug123) [inline .fantomasignore:1]: Why is this required?

File: src/Fantomas.Tests/StringTests.fs [hunk 2]
```diff
@@ -232,3 +232,23 @@ let ``newline in string`` () =
 
     List.length triviaNodes == 1
 """
+
+[<Test>]
+let ``double backslash in triple quote string`` () =
+    formatSourceString false "
+let a = \"\"\"\\\\\"\"\"
```
Comment (Smaug123) [inline src/Fantomas.Tests/StringTests.fs:239]: Oh my

File: src/Fantomas.Tests/TestHelpers.fs [hunk 3]
```diff
@@ -23,12 +23,7 @@ let private safeToIgnoreWarnings =
       "Identifiers containing '@' are reserved for use in F# code generation" ]
 
 let private isValidAndHasNoWarnings fileName source parsingOptions =
-    let allDefineOptions =
-        TokenParser.getOptimizedDefinesSets source
-        @ (TokenParser.getDefines source
-           |> List.map List.singleton)
-          @ [ [] ]
-        |> List.distinct
+    let allDefineOptions, _ = TokenParser.getDefines source
```
Comment (Smaug123) [inline src/Fantomas.Tests/TestHelpers.fs:26]: It was sad that there was a copy of this both here and in CodeFormatterImpl!

File: src/Fantomas/CodeFormatterImpl.fs [hunk 4]
```diff
@@ -359,29 +354,26 @@ let isValidFSharpCode (checker: FSharpChecker) (parsingOptions: FSharpParsingOpt
             let! ast = parse checker parsingOptions formatContext
 
             let isValid =
-                ast |> Array.map fst |> Array.forall isValidAST
+                ast
+                |> Array.map (fun (a, _, _) -> a)
```
Comment (Smaug123) [inline src/Fantomas/CodeFormatterImpl.fs:358]: Obvious allocation reduction: `ast |> Array.forall (fun (a, _, _) -> isValidAST a)`. But I imagine this is unlikely ever to matter.

File: src/Fantomas/CodeFormatterImpl.fs [hunk 5]
```diff
@@ -359,29 +354,26 @@ let isValidFSharpCode (checker: FSharpChecker) (parsingOptions: FSharpParsingOpt
...
         if String.IsNullOrWhiteSpace formatContext.Source
-        then None
-        else Some formatContext.Source
-
-    // Use '\n' as the new line delimiter consistently
-    // It would be easier for F# parser
-    let sourceCode = defaultArg input String.Empty
-    let normalizedSourceCode = String.normalizeNewLine sourceCode
+        then String.Empty, false
+        else formatContext.Source, true
```
Comment (Smaug123) [inline src/Fantomas/CodeFormatterImpl.fs:372]: This feels like a slightly broken domain. `sourceCode, hasInput` is really a `string option` that you're representing as a `string * bool`. This makes the `IsNullOrWhitespace` check on line 390 a bit harder to understand.

File: src/Fantomas/TokenParser.fs [hunk 6]
```diff
@@ -90,91 +90,189 @@ let private createHashToken lineNumber content fullMatchedLength offset =
...
-            FullMatchedLength = fullMatchedLength } }
+            FullMatchedLength = String.length content } }
 
-type private SourceCodeState =
+type SourceCodeState =
     | Normal
-    | InsideSingleQuoteString
-    | InsideTripleQuoteString
+    | InsideString
+    | InsideTripleQuoteString of int
```
Comment (Smaug123) [inline src/Fantomas/TokenParser.fs:98]: Could you put a docstring here indicating what the int is?

File: src/Fantomas/TokenParser.fs [hunk 7]
```diff
@@ -90,91 +90,189 @@ let private createHashToken lineNumber content fullMatchedLength offset =
...
+        let hashChar = '#'
+        let spaceChar = ' '
+        let isDoubleQuoteChar = (=) doubleQuoteChar
+        let isBackSlashChar = (=) backSlashChar
+        let isNoBackSlashChar v = v <> backSlashChar
+
+        let tripleQuotes =
+            (doubleQuoteChar, doubleQuoteChar, doubleQuoteChar)
+
+        let isOpenParenChar = (=) openParenChar
```
Comment (Smaug123) [inline src/Fantomas/TokenParser.fs:122]: Idle thought: could this all be represented better with active patterns? Instead of all these checks, you could have them all play nicely in a `match` statement.

File: src/Fantomas/TokenParser.fs [hunk 8]
```diff
@@ -90,91 +90,189 @@ let private createHashToken lineNumber content fullMatchedLength offset =
...
+
+                { t with
+                      LineNumber = lineNumber
+                      TokenInfo = info })
+            |> fun rest ->
+                (createHashToken lineNumber hashContent offset)
+                :: rest
+
+        let sourceLength = String.length sourceCode
+        let lastIndex = sourceLength - 3
```
Comment (Smaug123) [inline src/Fantomas/TokenParser.fs:155]: A comment here indicating what the `3` is referring to would be good

File: src/Fantomas/TokenParser.fs [hunk 9]
```diff
@@ -90,91 +90,189 @@ let private createHashToken lineNumber content fullMatchedLength offset =
...
+        let defines = ResizeArray()
+        let mutable currentState = Normal
+        let mutable newlineIndexes = ResizeArray()
```
Comment (Smaug123) [inline src/Fantomas/TokenParser.fs:158]: Pretty sure this won't need to be `mutable` (since the `ResizeArray` is already mutable).

If this were code in G-Research, I'd comment that this should be expressed immutably as a fold or by recursion; iterating over a list updating a state machine is a very suitable task for `fold`. But if imperative style is normal in Fantomas, then fine.

File: src/Fantomas/TokenParser.fs [hunk 10]
```diff
@@ -260,15 +354,24 @@ let getDefineExprs sourceCode =
...
           |> BoolExpr.mergeBoolExprs maxSteps
           |> List.map snd with
     | [] -> [ [] ]
     | xs -> xs
 
+let getDefines sourceCode =
+    let hashTokens = getTokenizedHashes sourceCode
+
+    let defineCombinations =
+        getOptimizedDefinesSets hashTokens @ (getDefinesWords  hashTokens |> List.map List.singleton) @ [ [] ]
```
Comment (Smaug123) [inline src/Fantomas/TokenParser.fs:370]: Hey, here's a formatting error (double space after `getDefinesWords`)! Sure we wanted that `.fantomasignore`? ;)

Comment (Smaug123) [inline src/Fantomas/TokenParser.fs:370]: Hey, here's a formatting error (double space after `getDefinesWords`)! Sure we wanted that `.fantomasignore`? ;)

**Review by Smaug123 — COMMENTED** [review]
I don't really understand this fully, but the shape of it certainly makes me think it looks better. (E.g. the fact that we parse out hash-tokens right at the beginning.)

The state machine is big and scary and I don't really feel like I understand it properly!

File: .fantomasignore (hunk 1 above)
Comment (nojaf) [reply .fantomasignore:1]: I discovered https://github.com/fsprojects/fantomas/issues/1211 so that is why I needed to exclude this from the `fantomas --check` we do inside the build script. A necessary evil until the next version drops.

**Comment by nojaf** [comment]
Thanks for the review @Smaug123, some very good remarks! I'll revisit this PR later this week.

File: src/Fantomas.Tests/StringTests.fs (hunk 2 above)
Comment (nojaf) [reply src/Fantomas.Tests/StringTests.fs:239]: yup 🙈

File: src/Fantomas.Tests/TestHelpers.fs (hunk 3 above)
Comment (nojaf) [reply src/Fantomas.Tests/TestHelpers.fs:26]: Indeed, although for the Benchmark it does not seem to have a noticeable impact that we now only call this once.
Unit tests do feel snappier though.

File: .fantomasignore (hunk 1 above)
Comment (nojaf) [reply .fantomasignore]: After refactoring I do have the `elif` without `else` constructs anymore so I was able to remove the `.fantomasignore`.

File: src/Fantomas/CodeFormatterImpl.fs (hunk 4 above)
Comment (nojaf) [reply src/Fantomas/CodeFormatterImpl.fs]: Good spot, never hesitate to mention things like this!

File: src/Fantomas/TokenParser.fs (hunk 7 above)
Comment (nojaf) [reply src/Fantomas/TokenParser.fs]: > Idle thought

You mean genius advice! The code to me is indeed more readable with the active patterns.

File: src/Fantomas/TokenParser.fs (hunk 9 above)
Comment (nojaf) [reply src/Fantomas/TokenParser.fs]: I went with an imperative style for no good reason. Refactoring to a `fold` was the right choice here.

**Comment by nojaf** [comment]
@Smaug123 I'm ready for round two 😄.

File: src/Fantomas/TokenParser.fs [hunk 11]
```diff
@@ -90,91 +89,213 @@ let private createHashToken lineNumber content fullMatchedLength offset =
...
+                |> Seq.take take
+                |> Seq.forall isSpace
+
+            if leadingCharactersBeforeLastNewlineAreSpaces then
+                let skip =
+                    if lastNewlineIdx = -1 then 0 else lastNewlineIdx + 1 // zero when the source starts with an #
+
+                let currentLine =
+                    sourceCode
+                    |> Seq.skip skip
```
Comment (Smaug123) [inline src/Fantomas/TokenParser.fs:183]: The use of `Seq.skip` always terrifies me, because it makes it very easy to write things which are quadratic in complexity; but strings are easy to seek around in, so that's fine.

**Review by Smaug123 — APPROVED** [review]
I can't say I'm 100% convinced of its correctness, but I don't see anything obviously wrong, and if no tests are broken then that's good enough for me!

**Comment by nojaf** [comment]
Sold!
